src/v/archival/archiver_service.cc

```diff
@@ -0,0 +1,999 @@
+
+        template<class T>
+        void on_entry(const T& event, state_machine_t& fsm) {
+            vlog(fsm._ctxlog.debug, "Enter {} from event {}", id, event);
+        }
+
+        template<class T>
+        void on_exit(const T& event, state_machine_t& fsm) {
+            vlog(fsm._ctxlog.debug, "Exit {} from event {}", id, event);
+        }
+    };
+
+    // List of states.
+    // All async states contains a future. State transition is triggered
+    // by the readiness of the future.
+    // Note on lifetimes. All states are created in c-tor and destroyed in
+    // d-tor. So they all exist at the same time inside the MSM. Because
+    // of that it's important to clear certain fields in every state when
+    // the state is exited. This is done by the 'clean' method that every
+    // state implements.
+
+    struct st_managed : state<managed_partition_state_t::managed> {
+        std::optional<ss::promise<>> trigger;
+        void clear() noexcept { trigger = std::nullopt; }
+    };
+    struct st_creating_async : state<managed_partition_state_t::creating> {
+        // ntp_archiver::start operation that runs in the background
+        std::optional<ssx::semaphore_units> reset_units;
+
+        void clear() noexcept { reset_units = std::nullopt; }
+    };
+    struct st_active : state<managed_partition_state_t::active> {
+        ss::lw_shared_ptr<ntp_archiver> archiver;
+
+        void clear() noexcept { archiver = nullptr; }
+    };
+    struct st_stopping_async : state<managed_partition_state_t::stopping> {
+        std::optional<ssx::semaphore_units> reset_units;
+
+        void clear() noexcept { reset_units = std::nullopt; }
+    };
+
+    // STM will be created with this state
+    using initial_state = st_managed;
+
+    // List of events
+    struct ev_leadership_acquired
+      : managed_partition_event_base<
+          managed_partition_event_t::leadership_acquired,
+          ev_leadership_acquired> {
+        model::term_id term;
+    };
+    struct ev_archiver_started
+      : managed_partition_event_base<
+          managed_partition_event_t::archiver_started,
+          ev_archiver_started> {
+        // started archiver
+        ss::lw_shared_ptr<ntp_archiver> archiver;
+    };
+    struct ev_archiver_failure
+      : managed_partition_event_base<
+          managed_partition_event_t::archiver_failure,
+          ev_archiver_failure> {
+        std::exception_ptr error;
+    };
+    struct ev_leadership_lost
+      : managed_partition_event_base<
+          managed_partition_event_t::leadership_lost,
+          ev_leadership_lost> {};
+    struct ev_archiver_stopped
+      : managed_partition_event_base<
+          managed_partition_event_t::archiver_stopped,
+          ev_archiver_stopped> {};
+    struct ev_shutdown
+      : managed_partition_event_base<
+          managed_partition_event_t::shutdown,
+          ev_shutdown> {};
+
+    // List of transitions.
+    struct managed_to_creating {
```

Comment:
should we prefix transitions with something like `tr_` for readability? similarily to st_ and ev_